File: dragonfly/gameover.py
import dfbase
import output.console
from enum import Enum

from PyQt5.QtXml import QDomElement, QDomNode

class ResultType(Enum):
	VICTORY = 0
	DEFEAT = 1

class GameOver:
	def __init__(self, dict: "dfbase.Dictionary") -> None:
		self.__dictionary = dict
		self.__gameOverMessage = "Game over"
		self.__gameOverStyle = ""
		self.__victoryMessage = "with VICTORY !"
		self.__victoryStyle = ""
		self.__defeatMessage = "with defeat"
		self.__defeatStyle = ""

	def run(self, result: ResultType, message: str) -> None:
		output.console.Console.println(self.__gameOverMessage, self.__gameOverStyle)
		if result == ResultType.VICTORY:
			output.console.Console.println(self.__victoryMessage, self.__victoryStyle)
		else:
			output.console.Console.println(self.__defeatMessage, self.__defeatStyle)
		output.console.Console.println(message)

	def load(self, element: QDomElement) -> None:
		for i in range(element.childNodes().count()):
			child = element.childNodes().at(i)
			if (child.nodeType() == QDomNode.ElementNode):
				if (child.nodeName() == "game-over-message"):
					self.__gameOverMessage, self.__gameOverStyle = \
					self.loadMessage(child)
				if (child.nodeName() == "victory-message"):
					self.__victoryMessage, self.__victoryStyle = \
					self.loadMessage(child)
				if (child.nodeName() == "defeat-message"):
					self.__defeatMessage, self.__defeatStyle = \
					self.loadMessage(child)

	def loadMessage(self, element: QDomElement):
		style = element.attribute("style")
		for i in range(element.childNodes().count()):
			child = element.childNodes().at(i)
			if (child.nodeType == QDomNode.TextNode):
				return child.toText().data().strip(), style
		return style, ""

File: base/gameover.js
import { DFMLElement } from "../dfml/js/main/element.js";
import { DFMLValue } from "../dfml/js/main/value.js";
import { Output } from "./output.js";
import { Utils } from "./utils.js";

export class ResultType {
  static VICTORY = 0;
  static DEFEAT = 1;
}

export class GameOver {
	constructor(dictionary) {
		this.dictionary = dictionary;
		this.gameOverMessage = "Game over";
		this.gameOverStyle = "";
		this.victoryMessage = "with VICTORY !";
		this.victoryStyle = "";
		this.defeatMessage = "with defeat";
		this.defeatStyle = "";
	}

	async run(result, action, conditions, responses) {

		for (const c of conditions) {
			if (!c.check(action)) return false;
		}

		Output.print(this.gameOverMessage, this.gameOverStyle);
		if (result === ResultType.VICTORY) {
			Output.print(this.victoryMessage, this.victoryStyle);
		} else {
			Output.print(this.defeatMessage, this.defeatStyle);
		}

		for (const r of responses) {
			await r.execute(action);
		}
	}

	load(node) {
		node.children.forEach((e) => {
			if (e.getElementType() === DFMLElement.NODE) {
				if (e.getName() === "game-over-message") {
					[this.gameOverMessage, this.gameOverStyle] = this.loadMessage(e);
				}
				if (e.getName() === "victory-message") {
					[this.victoryMessage, this.victoryStyle] = this.loadMessage(e);
				}
				if (e.getName() === "defeat-message") {
					[this.defeatMessage, this.defeatStyle] = this.loadMessage(e);
				}
			}
		});
	}

	loadMessage(node) {

		if (!Utils.expectedAttributes(node, "style")) return ;

		let style = "";
		if (node.hasAttr("style")) {
			style = node.getAttr("style").getValue();
		}

		node.children.forEach((e)=> {
			if (e.getElementType() === DFMLElement.DATA) {
				if (e.getValue().getType() === DFMLValue.STRING) {
					return [style, e.getValue().getValue()];
				}
			}
		});

		return [style, ""];
	}
}

